pkg/condition/condition.go

```diff
@@ -0,0 +1,109 @@
+package condition
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"time"
+
+	operators "github.com/operator-framework/api/pkg/operators/v2"
+	"github.com/pkg/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	patcher "github.com/openshift/windows-machine-config-operator/pkg/patch"
+)
+
+const (
+	upgradeableTrueReason   = "upgradeIsSafe"
+	upgradeableTrueMessage  = "The operator is safe for upgrade."
+	upgradeableFalseReason  = "upgradeIsNotSafe"
+	upgradeableFalseMessage = "The operator is currently processing sub-components."
+	// operatorConditionName is an environment variable set by OLM identifying the operator's OperatorCondition CR
+	OperatorConditionName = "OPERATOR_CONDITION_NAME"
+)
+
+// EnableOLMUpgrade is a wrapper that sets the Upgradeable condition to True to allow OLM to perform operator upgrades
+func EnableOLMUpgrade(c client.Client, watchNamespace string) error {
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionTrue, upgradeableTrueReason,
+		upgradeableTrueMessage)
+}
+
+// DisableOLMUpgrade is a wrapper that sets the Upgradeable condition to False to prevent operator upgrades by OLM
+func DisableOLMUpgrade(c client.Client, watchNamespace string) error {
+	return set(c, watchNamespace, operators.Upgradeable, meta.ConditionFalse, upgradeableFalseReason,
+		upgradeableFalseMessage)
+}
+
+// set sets the given condition with the given values within the associated OperatorCondition.
+// No-op if operator is not managed by OLM
+func set(c client.Client, ocNamespace, conditionType string, status meta.ConditionStatus,
+	reason, message string) error {
+	// Check if operator is OLM-managed
+	if ocName, present := os.LookupEnv(OperatorConditionName); present {
+		oc, err := get(c, ocNamespace, ocName)
+		if err != nil {
+			return err
+		}
+		if err = patch(c, oc, conditionType, status, reason, message); err != nil {
+			return err
+		}
+		// Treat condition setting as a lock mechanism -- ensure spec change is actually picked up
+		return waitForOperatorCondition(c, ocNamespace, ocName, conditionType, status)
+	}
+	return nil
+}
+
+// waitForOperatorCondition repeatedly checks if the OperatorCondition resource's Status has been updated
+func waitForOperatorCondition(c client.Client, ocNamespace, ocName, conditionType string,
```

Comment:
Are you really waiting for the OperatorCondition or for it to be updated? I suggest renaming the method to `waitForOperatorConditioUpdate` or even to `wait`.

I also don't understand the logic to wait and keep retrying the get(). Shouldn't you be retrying the patch()` which is what can fail?

Reply:
The purpose was that to check that after a successful patch, does the status actually get updated? If we don't wait then a controller start processing things while OLM still thinks upgrades are safe.

However, I see your point, we should retry the patch itself as well.